[PRA-93] [docs] Add experimental documentation on Kubeflow integration with Spark (#34)
canonical/charmed-kubeflow-documentation: docs/how-to/manage/integrate-spark.rst

.. _integrate_spark:

Integrate with Charmed Apache Spark
============================================

.. note::

>    This feature is currently experimental. This guide exists here for experimental purposes.

Reviewer: Suggestion: This feels a bit repetitive and doesn’t provide the concrete, actionable guidance I’d expect from a how-to guide. As a reader, I’m left wondering what to do with it: is it something I should avoid, or when is it actually a good idea to use this guide?

```suggestion
   This feature is experimental. We can't recommend using this guide in a production environment.
```

This guide describes how Charmed Kubeflow (CKF) and the Charmed Apache Spark can be integrated using `Juju`_. 
This integration enables running Spark jobs in Kubeflow notebooks and pipelines.

.. _kf_spark_requirements:

---------------------
Requirements
---------------------

- Minimum system requirements are at least 8 cores CPU processor, 64GB of RAM and 150GB of disk space.
- `MicroK8s`_ and Juju. See :ref:`CKF supported versions <supported_kubeflow_versions>` 
  for more details about compatible versions of `Kubeflow <https://www.kubeflow.org/docs/releases/>`_ and Juju.

> - CLI tools like ``charmcraft`` and ``tox``.

Author: Thanks, updated in the next commit.

command in the ``kubeflow`` model.

.. _deploy_data_kubeflow_integrator:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Deploy and configure Data-Kubeflow Integrator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Within the ``kubeflow`` model, deploy the Data-Kubeflow Integrator as follows:

.. code-block:: bash

   juju deploy -m kubeflow data-kubeflow-integrator --channel=1/edge

Set the ``profile`` and ``spark-service-account`` config in the Data-Kubeflow Integrator charm to specify the name 
of Kubeflow profile where Spark is supposed to be enabled and the name of the Spark service account that is supposed
to be created respectively. 

.. code-block:: bash

   juju config -m kubeflow data-kubeflow-integrator \
       profile=* \
       spark-service-account=spark

Note that the value of ``profile`` is set as ``*`` to enable Spark in all Kubeflow profiles. Refer to 
`this documentation <https://canonical-charmed-spark.readthedocs-hosted.com/main/how-to/manage-service-accounts/>`_ 
to read more about the Spark service accounts.

.. _integrate_data_kubeflow_integrator_with_integration_hub:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Integrate Data-Kubeflow Integrator with Spark Integration Hub

Author: The reason I'm quite reluctant to move the Juju relations section to "deploy  and configure data-kubeflow integrator" is because to me, configuring a charm is `juju config`, and adding relations is something distinct from the charm configuration.

Perhaps if you would suggest me a better wording, I can update it to that.

> Integrate Data-Kubeflow Integrator with Spark Integration Hub

Reviewer: I do like having a summary before the steps. But if we only have three steps, one per component, we might replace the whole fragment https://github.com/canonical/charmed-kubeflow-documentation/blob/spark-kf-docs/docs/how-to/manage/integrate-spark.rst?plain=1#L47-L56 with just a single sentence, stating that you need these three components to achieve the goal:

'To integrate Charmed Kubeflow with Charmed Apache Spark, you need three charms: Spark Integration Hub, Data-Kubeflow Integrator, and Resource Dispatcher.'

That's so much shorter and provides similar information in a less repetitive way.

This section of the guide assumes you have a working Juju K8s controller and the ``terraform`` and ``charmcraft`` CLI installed.

The steps to deploy the Charmed Kubeflow-Spark solution using Terraform is pretty similar to 
:ref:`Deploy Charmed Kubeflow using Terraform <install_terraform>`, the only difference being the Terraform module that is to be 
applied.

.. _fetch_kubeflow_spark_terraform_module:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Fetch the Charmed Kubeflow-Spark Terraform module
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Clone the ``charmed-kubeflow-solutions`` repository and change directory to the Kubeflow-Spark module as follows:

.. code-block:: bash

   git clone https://github.com/canonical/charmed-kubeflow-solutions
   cd charmed-kubeflow-solutions/modules/kubeflow-spark

.. _apply_terraform_module:

~~~~~~~~~~~~~~~~~~~~~~~~~~~
Apply the Terraform module
~~~~~~~~~~~~~~~~~~~~~~~~~~~

First of all, initialise Terraform. The following command downloads all the required 
`Terraform modules <https://developer.hashicorp.com/terraform/language/modules>`_ and installs the Terraform 

> `Juju provider <https://registry.terraform.io/providers/juju/juju/latest/docs>`_.

Reviewer: Suggestion: Let's use a colon here.

```suggestion
`Juju provider <https://registry.terraform.io/providers/juju/juju/latest/docs>`_:
```

.. code-block:: bash

   terraform init

Define credentials that will later be used to log into Kubeflow dashboard.

.. code-block:: bash

   DEX_USERNAME="your-username"
   DEX_PASSWORD="your-password"

Finally, deploy Charmed Kubeflow-Spark solution using Terraform as follows:

.. code-block:: bash

   terraform apply \
      -var dex_static_username=$DEX_USERNAME \
      -var dex_static_password=$DEX_PASSWORD \
      -var metacontroller_operator_revision=551 \

>       -var resource_dispatcher_revision=434

Reviewer 2: Let's keep them for now, and we can update these based on the outcome of [this](https://github.com/canonical/charmed-kubeflow-solutions/pull/102#discussion_r2742469264)

The command above:

* Creates a `Juju model <https://juju.is/docs/juju/model>`_ named ``kubeflow``.  
* Deploys CKF ``1.10/stable``.  
* Deploys charms like Spark Integration Hub, Data-Kubeflow Integrator and Resource Dispatcher that are necessary 
  for Spark integration.
* Configures `dex-auth <https://charmhub.io/dex-auth>`_ charm with a static user username and password.  
* Deploys a specific revision of `metacontroller-operator <https://charmhub.io/metacontroller-operator>`_ 
  and `resource-dispatcher <https://charmhub.io/resource-dispatcher>`_ charms because the changes necessary for 
  Spark integration aren't released to the stable channel yet.

Wait until the deployment is complete, and the ``terraform apply`` command returns.

> 

Reviewer: ```suggestion
```

.. _verify_kubeflow_spark_deployment:

~~~~~~~~~~~~~~~~~~~~~~~
Verify the deployment
~~~~~~~~~~~~~~~~~~~~~~~

As the first step, verify all charms are in ``active`` status by monitoring the Juju model:

.. code-block:: bash

   juju switch kubeflow

>    juju status --watch 1s

Reviewer: Suggestion: We tend to use `watch juju status` instead as the best practice (due to the potential memory leak in the one above). You can use the `-n` option for the interval.

Author: Thanks, fixed on 48bd701.

.. note::

   This may take up to some minutes, depending on the cluster's node specifications.

For additional validation, Charmed Kubeflow-Spark User Acceptance Tests (UAT) can be run on top of this deployment
to verify that the deployment was indeed successful and that Spark is enabled in the Kubeflow notebooks and pipelines.

To run the UAT on top of the deployment, first fetch the UAT tests from the ``charmed-kubeflow-uats`` repository:

.. code-block:: bash

   git clone https://github.com/canonical/charmed-kubeflow-uats.git
   cd charmed-kubeflow-uats

Now run the following command to run the UAT:

.. code-block:: bash

   tox -e spark-remote -- \
       --test-image ghcr.io/canonical/charmed-spark-jupyterlab:3.5-22.04_edge

> This will run the tests to verify that Spark is enabled in both Kubeflow notebooks and Kubeflow pipeline steps.

Reviewer: Suggestion: Let's add a short reference to what to expect as a success and how much time, roughly, it could take. As a quality of life improvement. It should not be too specific to simplify maintenance of the documentation.

---------------------

> Access CKF dashboard

Author: I get your point. Perhaps it's better to rephrase the section as `Run Spark Jobs` and then give a one liner description about what the objective next would be, giving the user the link to follow to do that. A little rephrasing would be better.

> Access CKF dashboard

Author: The thing is, the instructions to get the IP address and access the dashboard is already covered elsewhere, and thus I didn't want to repeat the the same here. What do you suggest we change this heading into?

Maybe "Further steps"?
